V2 migration

emmanueldenloye opened the pull request.

emmanueldenloye committed:
init commit

emmanueldenloye committed:
1st cut of migration

If we find old noderefs (e.g. us1.testnet.chainweb.com), we remove them
and put the correct uri (api.testnet.chainweb.com)

jmininger commented:
You should make sure that the tests in `frontend/tests/Frontend/VersionedStore*` are changed over from V1 to V2 (or that V2 is added?)

jmininger commented on frontend/src/Frontend/VersionedStore/V2.hs:
```diff
@@ -0,0 +1,216 @@
+    newNetworks = (\nets -> StoreFrontend_Network_Networks :=> Identity (convertNodeRefs $ V0.unNetworkMap $ runIdentity nets))
+      <$> DMap.lookup V0.StoreNetwork_Networks v0
+
+    -- This will regenerate the missing key. Desktop will recover the key with
```
This is something we really need to look out for with the upcoming chainweaver changes. Might be ok in the immediate future, but we can't copy the same code over for a V2 --> V3 if this ever happens in the future, because then we would risk wiping out users' keys

jmininger commented on frontend/src/Frontend/VersionedStore/V2.hs:
```diff
@@ -0,0 +1,216 @@
+
+deriving instance Show (StoreFrontend key a)
+
+upgradeFromV0 :: (Monad m, HasCrypto key m) => DMap (V0.StoreFrontend key) Identity -> m (DMap (StoreFrontend key) Identity)
```
Might it make more sense to just compose `V1.upgradeFromV0` and `V2.upgradeFromV1` so that we aren't stuck copying code over for each migration and having to make sure that we have all of the "previous stuff" from the intermediary migration as well as all of the new migration stuff?

mightybyte replied:
Hmmmm, interesting point.  I think we probably should compose the migrations otherwise the logical conclusion just won't be scalable.  That's typically the way I've done it with DB migrations.

jmininger commented on frontend/src/Frontend/VersionedStore.hs:
```diff
@@ -161,3 +168,4 @@ versionedStorage = VersionedStorage
           liftIO $ Api._transactionLogger_rotateLogFile txLogger
           pure ()
         Just (StoreFrontendVersion_1 :=> _) -> pure ()
+        Just (StoreFrontendVersion_2 :=> _) -> pure ()
```
Reminder to fix this

emmanueldenloye committed:
compose migrations

emmanueldenloye committed:
bump version on restoreBackup

emmanueldenloye committed:
don't care about power users

just replace the entries for the network names "Testnet" and "Mainnet"

emmanueldenloye committed:
add V1->V2 test, fix V0->V2 test

also fixed upgrade procedure

emmanueldenloye committed:
actually run both upgrade tests

emmanueldenloye committed:
mocking new test cases

these new test cases will investigate how chainweaver
recovers from a migration that fails at some point before
completion.

emmanueldenloye committed:
add OverloadedStrings pragma

emmanueldenloye committed:
finished previously mocked tests (last commit)

mightybyte commented on frontend/src/Frontend/Storage/Class.hs:
```diff
@@ -169,6 +169,7 @@ restoreLocalStorage
   :: forall storeKeys m
   . ( HasStorage m
     , Monad m
+    , MonadIO m
```
What was the reason for these new `MonadIO` constraints?  Is it left over from debugging?

emmanueldenloye replied:
Yeah,  I'm pretty sure that's the reason. I'll take them out.

mightybyte commented on frontend/tests/Frontend/VersionedStoreSpec.hs:
```diff
@@ -292,7 +374,169 @@ test_v0ToV1Upgrade = testCaseSteps "V0 to V1 Upgrade" $ \step -> do
   where
     path = "tests" </> "Frontend" </> "VersionedStoreSpec.files" </> "V0"
 
+fail_test_v0ToV2Upgrade :: FailStorageState -> TestTree
```
If we use compositional migrations as mentioned above, we might be able to get this test to become more like a property, i.e.

```
migrate a (a+2) == (migrate a (a+1) >> migrate (a+1) (a+2))
```

Not sure if that will be too invasive of a change though.

emmanueldenloye replied:
I actually really like this idea. Also adding new tests for new versions will be a lot easier to make.

mightybyte commented on frontend/src/Frontend/VersionedStore/V2.hs:
```diff
@@ -0,0 +1,218 @@
+convertNodeRefs :: Map NetworkName [NodeRef] -> Map NetworkName [NodeRef]
+convertNodeRefs = Map.mapWithKey migrate
+  where
+    migrate = \case
```
Hmmmm...I'm contemplating the possibility of doing the replacement no matter what the network name is.  I think I've seen instances where it's lowercase "mainnet" rather than "Mainnet".  We would have to keep track of which of the node refs were actually found to know which new ref to add.  But that would work for these other cases.

jmininger commented on frontend/src/Frontend/VersionedStore/V2.hs:
```diff
@@ -0,0 +1,216 @@
+
+upgradeFromV0 :: (Monad m, HasCrypto key m) => DMap (V0.StoreFrontend key) Identity -> m (DMap (StoreFrontend key) Identity)
+upgradeFromV0 v0 = do
+  (newKeysList, newAccountStorage) <- foldMapM splitOldKey oldKeysList
```
`splitOldKey` really concerns me, for the web version. Its relying on the keys' index in a list, and then using that index to "regenerate" them instead of saving /copying them somewhere. This code is so easy to accidentally copy over, for, say the next migration. We should either rewrite it or at the very least use CPP to make it a no-op on ghcjs. Not quite sure its in the scope of this task, but it could so easily go really bad really quickly if it isn't taken care of imo

emmanueldenloye committed:
take out unnecessary MonadIO constraints

emmanueldenloye committed:
unconditionally alter refs for networks in local storage

In essence, we don't care what networks the user has either created,
modified or left alone, we filter out undesired noderefs in that network name.

emmanueldenloye committed:
create simplified version of "These" (Cheese)

It is equivalent to "These" except that we take out the
"This" constructor. With it gone, we still get the same applicative
operations from "These" (sans "This") but we don't have to account for
the "This" constructor in situations where we know it is impossible to
produce it.

jmininger commented on frontend/src/Frontend/VersionedStore/V2.hs:
```diff
@@ -152,16 +155,19 @@ fromMultiSet = ($ []) . Map.foldrWithKey (\k i -> (.) (dlrep k i)) id
 convertNodeRefs :: Map NetworkName [NodeRef] -> Map NetworkName [NodeRef]
-convertNodeRefs = Map.mapWithKey migrate
+convertNodeRefs = fmap migrate
   where
```
Is the nested `where` necessary ?

emmanueldenloye committed:
Use an even simpler datatype for processing the map merge

The datatype we really want is something equivalent to Either () a but
with a different Applicative instance.

emilypi commented on frontend/src/Frontend/VersionedStore/V2.hs:
```diff
@@ -153,32 +153,37 @@ fromMultiSet = ($ []) . Map.foldrWithKey (\k i -> (.) (dlrep k i)) id
+-- the "This" constructor in situations where we know it is impossible to
+-- produce it. This is also morally equivalent to Either () b with a
+-- different applicative instance
+data Deez b = Dis b | Dat b
```
slightly misleading comment. 

```haskell
These a b 
~ a + b + a*b
```

But

```haskell
These () a 
~ 1 + a + 1*a 
~ 1 + a + a 
~ Maybe (Either a a)
```

and 

```haskell
Either () b 
~ 1 + b
~ Maybe b
```

You want a newtype for `Either a a` here to get the modified `Applicative`.

emmanueldenloye committed:
another simple refactor

make "Deez" a newtype

emmanueldenloye committed:
for JM
